File: backend/src/routes/groups.js
import express from 'express';
import { Group } from '../models/Group.js';
import { authenticateUser } from '../middleware/auth.js';

const router = express.Router();

// Get all groups for authenticated user
router.get('/', authenticateUser, async (req, res) => {
  try {
    const groups = await Group.find({ userId: req.user.uid })
      .sort({ createdAt: -1 });

    res.json({
      success: true,
      data: groups.map(g => ({
        id: g.id,
        name: g.name,
        members: g.members,
        expenses: g.expenses,
        createdAt: g.createdAt,
      })),
    });
  } catch (error) {
    console.error('Get groups error:', error);
    res.status(500).json({
      success: false,
      message: 'Error fetching groups',
    });
  }
});

// Get single group
router.get('/:groupId', authenticateUser, async (req, res) => {
  try {
    const group = await Group.findOne({
      id: req.params.groupId,
      userId: req.user.uid,
    });

    if (!group) {
      return res.status(404).json({
        success: false,
        message: 'Group not found',
      });
    }

    res.json({
      success: true,
      data: {
        id: group.id,
        name: group.name,
        members: group.members,
        expenses: group.expenses,
        createdAt: group.createdAt,
      },
    });
  } catch (error) {
    console.error('Get group error:', error);
    res.status(500).json({
      success: false,
      message: 'Error fetching group',
    });
  }
});

// Create new group
router.post('/', authenticateUser, async (req, res) => {
  try {
    console.log('Create group payload:', JSON.stringify(req.body));
    const { id, name, members } = req.body;

    if (!name || !members || members.length < 2) {
      return res.status(400).json({
        success: false,
        message: 'Group name and at least 2 members are required',
      });
    }

    const group = await Group.create({
      id: id || Date.now().toString(),
      name,
      members,
      expenses: [],
      userId: req.user.uid,
    });

    res.status(201).json({
      success: true,
      data: {
        id: group.id,
        name: group.name,
        members: group.members,
        expenses: group.expenses,
        createdAt: group.createdAt,
      },
    });
  } catch (error) {
    console.error('Create group error:', error);
    res.status(500).json({
      success: false,
      message: 'Error creating group',
    });
  }
});

// Update group
router.put('/:groupId', authenticateUser, async (req, res) => {
  try {
    const { name, members, expenses } = req.body;

    const group = await Group.findOneAndUpdate(
      { id: req.params.groupId, userId: req.user.uid },
      { name, members, expenses },
      { new: true }
    );

    if (!group) {
      return res.status(404).json({
        success: false,
        message: 'Group not found',
      });
    }

    res.json({
      success: true,
      data: {
        id: group.id,
        name: group.name,
        members: group.members,
        expenses: group.expenses,
        createdAt: group.createdAt,
      },
    });
  } catch (error) {
    console.error('Update group error:', error);
    res.status(500).json({
      success: false,
      message: 'Error updating group',
    });
  }
});

// Delete group
router.delete('/:groupId', authenticateUser, async (req, res) => {
  try {
    const group = await Group.findOneAndDelete({
      id: req.params.groupId,
      userId: req.user.uid,
    });

    if (!group) {
      return res.status(404).json({
        success: false,
        message: 'Group not found',
      });
    }

    res.json({
      success: true,
      message: 'Group deleted successfully',
    });
  } catch (error) {
    console.error('Delete group error:', error);
    res.status(500).json({
      success: false,
      message: 'Error deleting group',
    });
  }
});

// Add expense to group
router.post('/:groupId/expenses', authenticateUser, async (req, res) => {
  try {
    const { id, description, amount, paidBy, participants } = req.body;

    if (!description || !amount || !paidBy || !participants) {
      return res.status(400).json({
        success: false,
        message: 'All expense fields are required',
      });
    }

    const group = await Group.findOne({
      id: req.params.groupId,
      userId: req.user.uid,
    });

    if (!group) {
      return res.status(404).json({
        success: false,
        message: 'Group not found',
      });
    }

    group.expenses.push({
      id: id || Date.now().toString(),
      description,
      amount,
      paidBy,
      participants,
      date: new Date(),
    });

    await group.save();

    res.status(201).json({
      success: true,
      data: group.expenses[group.expenses.length - 1],
    });
  } catch (error) {
    console.error('Add expense error:', error);
    res.status(500).json({
      success: false,
      message: 'Error adding expense',
    });
  }
});

// Delete expense from group
router.delete('/:groupId/expenses/:expenseId', authenticateUser, async (req, res) => {
  try {
    const group = await Group.findOne({
      id: req.params.groupId,
      userId: req.user.uid,
    });

    if (!group) {
      return res.status(404).json({
        success: false,
        message: 'Group not found',
      });
    }

    group.expenses = group.expenses.filter(exp => exp.id !== req.params.expenseId);
    await group.save();

    res.json({
      success: true,
      message: 'Expense deleted successfully',
    });
  } catch (error) {
    console.error('Delete expense error:', error);
    res.status(500).json({
      success: false,
      message: 'Error deleting expense',
    });
  }
});

export default router;

File: backend/routes/groups.go
package routes

import (
	"context"
	"split-it/backend/config"
	"split-it/backend/middleware"
	"split-it/backend/models"
	"strconv"
	"time"

	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// SetupGroupRoutes configures group-related routes
func SetupGroupRoutes(app *fiber.App) {
	groups := app.Group("/api/groups", middleware.AuthenticateUser)

	// Group CRUD operations
	groups.Get("/", getAllGroups)
	groups.Get("/:groupId", getGroup)
	groups.Post("/", createGroup)
	groups.Put("/:groupId", updateGroup)
	groups.Delete("/:groupId", deleteGroup)

	// Expense operations
	groups.Post("/:groupId/expenses", addExpense)
	groups.Delete("/:groupId/expenses/:expenseId", deleteExpense)
}

func getAllGroups(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	opts := options.Find().SetSort(bson.D{{Key: "createdAt", Value: -1}})
	cursor, err := collection.Find(ctx, bson.M{"userId": user.UID}, opts)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error fetching groups",
		})
	}
	defer cursor.Close(ctx)

	var groups []models.Group
	if err = cursor.All(ctx, &groups); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error decoding groups",
		})
	}

	// Transform to response format
	response := make([]models.GroupResponse, len(groups))
	for i, g := range groups {
		response[i] = models.GroupResponse{
			ID:        g.GroupID,
			Name:      g.Name,
			Members:   g.Members,
			Expenses:  g.Expenses,
			CreatedAt: g.CreatedAt,
		}
	}

	return c.JSON(fiber.Map{
		"success": true,
		"data":    response,
	})
}

func getGroup(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	groupId := c.Params("groupId")

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var group models.Group
	err := collection.FindOne(ctx, bson.M{
		"id":     groupId,
		"userId": user.UID,
	}).Decode(&group)

	if err == mongo.ErrNoDocuments {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Group not found",
		})
	} else if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error fetching group",
		})
	}

	return c.JSON(fiber.Map{
		"success": true,
		"data": models.GroupResponse{
			ID:        group.GroupID,
			Name:      group.Name,
			Members:   group.Members,
			Expenses:  group.Expenses,
			CreatedAt: group.CreatedAt,
		},
	})
}

func createGroup(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	var body struct {
		ID      string          `json:"id"`
		Name    string          `json:"name"`
		Members []models.Member `json:"members"`
	}

	if err := c.BodyParser(&body); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
		})
	}

	if body.Name == "" || len(body.Members) < 2 {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Group name and at least 2 members are required",
		})
	}

	// Generate ID if not provided
	groupID := body.ID
	if groupID == "" {
		groupID = strconv.FormatInt(time.Now().UnixNano(), 10)
	}

	newGroup := models.Group{
		ID:        primitive.NewObjectID(),
		GroupID:   groupID,
		Name:      body.Name,
		Members:   body.Members,
		Expenses:  []models.Expense{},
		UserID:    user.UID,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	_, err := collection.InsertOne(ctx, newGroup)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error creating group",
		})
	}

	return c.Status(fiber.StatusCreated).JSON(fiber.Map{
		"success": true,
		"data": models.GroupResponse{
			ID:        newGroup.GroupID,
			Name:      newGroup.Name,
			Members:   newGroup.Members,
			Expenses:  newGroup.Expenses,
			CreatedAt: newGroup.CreatedAt,
		},
	})
}

func updateGroup(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	groupId := c.Params("groupId")

	var body struct {
		Name     string           `json:"name"`
		Members  []models.Member  `json:"members"`
		Expenses []models.Expense `json:"expenses"`
	}

	if err := c.BodyParser(&body); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
		})
	}

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	update := bson.M{
		"$set": bson.M{
			"name":      body.Name,
			"members":   body.Members,
			"expenses":  body.Expenses,
			"updatedAt": time.Now(),
		},
	}

	var updatedGroup models.Group
	err := collection.FindOneAndUpdate(
		ctx,
		bson.M{"id": groupId, "userId": user.UID},
		update,
		options.FindOneAndUpdate().SetReturnDocument(options.After),
	).Decode(&updatedGroup)

	if err == mongo.ErrNoDocuments {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Group not found",
		})
	} else if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error updating group",
		})
	}

	return c.JSON(fiber.Map{
		"success": true,
		"data": models.GroupResponse{
			ID:        updatedGroup.GroupID,
			Name:      updatedGroup.Name,
			Members:   updatedGroup.Members,
			Expenses:  updatedGroup.Expenses,
			CreatedAt: updatedGroup.CreatedAt,
		},
	})
}

func deleteGroup(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	groupId := c.Params("groupId")

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	result, err := collection.DeleteOne(ctx, bson.M{
		"id":     groupId,
		"userId": user.UID,
	})

	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error deleting group",
		})
	}

	if result.DeletedCount == 0 {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Group not found",
		})
	}

	return c.JSON(fiber.Map{
		"success": true,
		"message": "Group deleted successfully",
	})
}

func addExpense(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	groupId := c.Params("groupId")

	var body struct {
		ID           string   `json:"id"`
		Description  string   `json:"description"`
		Amount       float64  `json:"amount"`
		PaidBy       string   `json:"paidBy"`
		Participants []string `json:"participants"`
	}

	if err := c.BodyParser(&body); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
		})
	}

	if body.Description == "" || body.Amount == 0 || body.PaidBy == "" || len(body.Participants) == 0 {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "All expense fields are required",
		})
	}

	// Generate ID if not provided
	expenseID := body.ID
	if expenseID == "" {
		expenseID = strconv.FormatInt(time.Now().UnixNano(), 10)
	}

	newExpense := models.Expense{
		ID:           expenseID,
		Description:  body.Description,
		Amount:       body.Amount,
		PaidBy:       body.PaidBy,
		Participants: body.Participants,
		Date:         time.Now(),
	}

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	update := bson.M{
		"$push": bson.M{"expenses": newExpense},
		"$set":  bson.M{"updatedAt": time.Now()},
	}

	var updatedGroup models.Group
	err := collection.FindOneAndUpdate(
		ctx,
		bson.M{"id": groupId, "userId": user.UID},
		update,
		options.FindOneAndUpdate().SetReturnDocument(options.After),
	).Decode(&updatedGroup)

	if err == mongo.ErrNoDocuments {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Group not found",
		})
	} else if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error adding expense",
		})
	}

	// Return the newly added expense
	return c.Status(fiber.StatusCreated).JSON(fiber.Map{
		"success": true,
		"data":    newExpense,
	})
}

func deleteExpense(c *fiber.Ctx) error {
	user := middleware.GetUserFromContext(c)
	if user == nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"success": false,
			"message": "Unauthorized",
		})
	}

	groupId := c.Params("groupId")
	expenseId := c.Params("expenseId")

	db := config.GetDB()
	collection := db.Collection("groups")

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	update := bson.M{
		"$pull": bson.M{"expenses": bson.M{"id": expenseId}},
		"$set":  bson.M{"updatedAt": time.Now()},
	}

	result, err := collection.UpdateOne(
		ctx,
		bson.M{"id": groupId, "userId": user.UID},
		update,
	)

	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error deleting expense",
		})
	}

	if result.MatchedCount == 0 {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Group not found",
		})
	}

	return c.JSON(fiber.Map{
		"success": true,
		"message": "Expense deleted successfully",
	})
}
